server/raft.go

```diff
@@ -1758,6 +1766,13 @@ func (n *raft) shutdown(shouldDelete bool) {
 		n.c = nil
 	}
 
+	// Wait for goroutines to shut down before trying to clean up
+	// the log below, otherwise we might end up deleting the store
+	// from underneath running goroutines.
+	n.Unlock()
```

Comment:
We have these patterns for sure elsewhere but I try hard to avoid them if at all possible.